presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeletedRows.java

```diff
@@ -0,0 +1,251 @@
+    private final HdfsEnvironment hdfsEnvironment;
+
+    private Set<RowId> deletedRows;
+
+    public OrcDeletedRows(
+            String sourceFileName,
+            Optional<DeleteDeltaLocations> deleteDeltaLocations,
+            OrcDeletedDeltaPageSourceFactory pageSourceFactory,
+            String sessionUser,
+            Configuration configuration,
+            HdfsEnvironment hdfsEnvironment)
+    {
+        this.sourceFileName = sourceFileName;
+        this.pageSourceFactory = pageSourceFactory;
+        this.sessionUser = sessionUser;
+        this.configuration = configuration;
+        this.hdfsEnvironment = hdfsEnvironment;
+        if (deleteDeltaLocations.isPresent()) {
+            this.deleteDeltaLocations = deleteDeltaLocations.get();
+        }
+        else {
+            this.deletedRows = ImmutableSet.of();
+            this.deleteDeltaLocations = null;
+        }
+    }
+
+    public MaskDeletedRowsFunction getMaskDeletedRowsFunction(Page pageRowIds)
+    {
+        return new MaskDeletedRowsFunction(pageRowIds);
+    }
+
+    @NotThreadSafe
+    public class MaskDeletedRowsFunction
+            implements Function<Block, Block>
+    {
+        private Page pageRowIds;
+        private int positionCount;
+        private int[] validPositions;
+
+        public MaskDeletedRowsFunction(Page pageRowIds)
+        {
+            this.pageRowIds = pageRowIds;
+        }
+
+        public int getPositionCount()
+        {
+            if (pageRowIds != null) {
+                loadValidPositions();
+            }
+
+            return positionCount;
+        }
+
+        @Override
+        public Block apply(Block block)
+        {
+            if (pageRowIds != null) {
+                loadValidPositions();
+            }
+
+            if (positionCount == block.getPositionCount()) {
+                return block;
+            }
+            return new DictionaryBlock(positionCount, block, validPositions);
+        }
+
+        private void loadValidPositions()
+        {
+            Set<RowId> deletedRows = getDeletedRows();
+            if (deletedRows.isEmpty()) {
+                this.positionCount = pageRowIds.getPositionCount();
+                this.pageRowIds = null;
+                return;
+            }
+
+            int[] validPositions = new int[pageRowIds.getPositionCount()];
+            int validPositionsIndex = 0;
+            for (int position = 0; position < pageRowIds.getPositionCount(); position++) {
+                if (!deletedRows.contains(new RowId(pageRowIds, position))) {
+                    validPositions[validPositionsIndex++] = position;
+                }
+            }
+            this.positionCount = validPositionsIndex;
+            this.validPositions = validPositions;
+            this.pageRowIds = null;
+        }
+    }
+
+    private Set<RowId> getDeletedRows()
+    {
+        if (deletedRows != null) {
+            return deletedRows;
```

Comment:
I think `getDeletedRows()` is being called concurrently and the class is not thread-safe.

Reply:
I am wrong. This is confined to a thread. It means we're re-reading delete deltas for every split.
Maybe we could even benefit from that -- since we know base files data range (writer id, etc.), we could use this as a filter when reading delete deltas.
@dain  thoughts? Can you create appropriate issue about this?

Reply:
I believe that is an optimization outside of this PR.  You could make the same argument for file footer.  I'm not convinced that there is a big win here.  Once you get more than a few machines, the hit rate will fall as the splits are spread across machines.